pkg/asset/installconfig/openstack/validation/platform.go

```diff
@@ -37,26 +37,74 @@ func ValidatePlatform(p *openstack.Platform, n *types.Networking, ci *CloudInfo)
+	networkID := ""
+	hasIPv4Subnet := false
+	hasIPv6Subnet := false
+	if p.ControlPlanePort != nil {
 		if len(p.ExternalDNS) > 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("externalDNS"), p.ExternalDNS, "externalDNS is set, externalDNS is not supported when machinesSubnet is set"))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("externalDNS"), p.ExternalDNS, "externalDNS is set, externalDNS is not supported when ControlPlanePort is set"))
 		}
-		if ci.MachinesSubnet == nil {
-			allErrs = append(allErrs, field.NotFound(fldPath.Child("machinesSubnet"), p.MachinesSubnet))
-		} else if !validUUIDv4(p.MachinesSubnet) {
-			allErrs = append(allErrs, field.InternalError(fldPath.Child("machinesSubnet"), errors.New("invalid subnet ID")))
-		} else {
-			if n.MachineNetwork[0].CIDR.String() != ci.MachinesSubnet.CIDR {
-				allErrs = append(allErrs, field.InternalError(fldPath.Child("machinesSubnet"), fmt.Errorf("the first CIDR in machineNetwork, %s, doesn't match the CIDR of the machineSubnet, %s", n.MachineNetwork[0].CIDR.String(), ci.MachinesSubnet.CIDR)))
+		for _, fixedIP := range p.ControlPlanePort.FixedIPs {
+			subnet := getSubnet(ci.ControlPlanePortSubnets, fixedIP.Subnet.ID, fixedIP.Subnet.Name)
+			if subnet == nil {
+				subnetDetail := fixedIP.Subnet.ID
+				if subnetDetail == "" {
+					subnetDetail = fixedIP.Subnet.Name
+				}
+				allErrs = append(allErrs, field.NotFound(fldPath.Child("controlPlanePort").Child("fixedIPs"), subnetDetail))
+			} else {
+				if subnet.IPVersion == 6 {
+					hasIPv6Subnet = true
+				} else {
+					hasIPv4Subnet = true
+				}
+				if !isSubnetCIDRInMachineNetworkCIDR(n.MachineNetwork, subnet) {
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("controlPlanePort").Child("fixedIPs"), subnet.CIDR, "controlPlanePort CIDR does not match machineNetwork"))
+				}
+				if networkID != "" && networkID != subnet.NetworkID {
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("controlPlanePort").Child("fixedIPs"), subnet.NetworkID, "fixedIPs subnets must be on the same Network"))
+				}
+				networkID = subnet.NetworkID
 			}
 		}
+		if !hasIPv4Subnet && hasIPv6Subnet {
+			allErrs = append(allErrs, field.InternalError(fldPath.Child("controlPlanePort").Child("fixedIPs"), fmt.Errorf("one IPv4 subnet must be specified")))
+		} else if hasIPv4Subnet && !hasIPv6Subnet && len(p.ControlPlanePort.FixedIPs) == 2 {
+			allErrs = append(allErrs, field.InternalError(fldPath.Child("controlPlanePort").Child("fixedIPs"), fmt.Errorf("multiple IPv4 subnets is not supported")))
+		}
+		controlPlaneNetwork := p.ControlPlanePort.Network
+		if controlPlaneNetwork.ID != "" && controlPlaneNetwork.ID != networkID {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("controlPlanePort").Child("network"), controlPlaneNetwork.ID, "network must contain subnets"))
+		} else if controlPlaneNetwork.Name != "" && ci.ControlPlanePortNetwork.ID != networkID {
```

Comment:
This check is unclear. Could you please add a comment explaining what is being validated here? 
Also, if controlPlaneNetwork.ID is set, you will not validate controlPlaneNetwork.Name even if it is set? Is that intentional?

Reply:
`ControlPlanePortNetwork ` retrieves the network from openstack based on the name only. If the ID of the retrieved network does not match the ID of the network of the subnets it means the network provided is a wrong one.
I will add a comment explaining that

Reply:
I ended up reshaping it a bit.